src/testStatusCodeLensProvider.ts

```diff
@@ -0,0 +1,78 @@
+"use strict";
+import { CancellationToken, CodeLens, CodeLensProvider, commands, Disposable, Event, EventEmitter, Range, SymbolInformation, SymbolKind, TextDocument, workspace } from "vscode";
+import { TestResult } from "./testResult";
+import { TestResultsFile } from "./testResultsFile";
+import { TestStatusCodeLens } from "./testStatusCodeLens";
+
+export class TestStatusCodeLensProvider implements CodeLensProvider {
+    private disposables: Disposable[] = [];
+    private enabled: boolean;
+    private onDidChangeCodeLensesEmitter = new EventEmitter<void>();
+
+    // Store everything in a map so we can remember old tests results for the
+    // scenario where a single test is ran. If the test no longer exists in
+    // code it will never be mapped to the symbol, so no harm (though there is
+    // a memory impact)
+    private testResults = new Map<string, TestResult>();
+
+    public constructor(testResultFile: TestResultsFile) {
+        this.checkEnabledOption();
+
+        this.disposables.push(
+            testResultFile.onNewResults(this.addTestResults, this));
+
+        this.disposables.push(
+            workspace.onDidChangeConfiguration(this.checkEnabledOption, this));
+    }
+
+    public dispose() {
+        while (this.disposables.length) {
+            this.disposables.pop().dispose();
+        }
+    }
+
+    public get onDidChangeCodeLenses(): Event<void> {
+        return this.onDidChangeCodeLensesEmitter.event;
+    }
+
+    public provideCodeLenses(document: TextDocument, token: CancellationToken): CodeLens[] | Thenable<CodeLens[]> {
+        if (!this.enabled) {
+            return [];
+        }
+
+        const results = this.testResults;
+        return commands.executeCommand<SymbolInformation[]>("vscode.executeDocumentSymbolProvider", document.uri)
+            .then((symbols) => {
+                const mapped: CodeLens[] = [];
+                for (const symbol of symbols.filter((x) => x.kind === SymbolKind.Method)) {
+                    const fullName = symbol.containerName + "." + symbol.name;
+                    for (const result of results.values()) {
+                        if (result.testName.endsWith(fullName)) {
```

Comment:
There is a corner case:
![image](https://user-images.githubusercontent.com/1050213/32825657-b92e7a1a-ca20-11e7-89b9-3c712b43e50c.png)

Reply:
The symbol provider doesn't return anything for the namespace and the container of the class is empty, so would need to look at perhaps using omni-sharp directly? There's actually an issue with the test discovery with NUnit that only lists the method names, so if two tests exist with the same method name in different classes, NUnit only returns one of them.

Not sure how you want to approach this one? Accept it as a corner case until the extension uses omni-sharp to parse the files to discover the unit test methods?

Reply:
I think this corner case could be accepted.

Reply:
If you have interest to look at omni-sharp or other approach, it is warmly welcome! You could send PR in the future.